### PR TITLE
fix(stepfunctions): no validation on state machine name

### LINKED ISSUE #13289: (aws-stepfunctions): Validate StateMachine name
There are a couple of constraints on StateMachine name. Those should be validated when creating the CloudFormation template.

Here are the constraints: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-stepfunctions-statemachine.html#cfn-stepfunctions-statemachine-statemachinename

In addition there is also length constraint, which is not specified in Cfn documentation: https://docs.aws.amazon.com/step-functions/latest/apireference/API_CreateStateMachine.html#StepFunctions-CreateStateMachine-request-name

### Use Case

Catching errors before deploying

### Proposed Solution
```
if (!validName(name)) {
throw Error
}
```

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by jonnekaunisto**
closes #13289

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts [hunk 1]
```diff
@@ -426,6 +430,15 @@ export class StateMachine extends StateMachineBase {
     this.role.addToPrincipalPolicy(statement);
   }
 
+  private validateStateMachineName(stateMachineName: string) {
+    if (stateMachineName.length < 1 || stateMachineName.length > 80) {
+      throw new Error('StateMachine name length must be between 1 and 80 characters');
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:435]: also provide the input that was supplied when throwing this error so users have a better idea around why validation failed.

```suggestion
      throw new Error(`State Machine name must be between 1 and 80 characters. received: ${stateMachineName}`);
```

File: packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts [hunk 2]
```diff
@@ -426,6 +430,15 @@ export class StateMachine extends StateMachineBase {
...
+    }
+    if (stateMachineName.match('[0-9a-zA-Z\_\-]+')?.length != 1) {
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:437]: can this be simplified to just use a regex with the match.

```suggestion
    if (stateMachineName.match('/^[0-9a-zA-Z\_\-]+$/')) {
```

not 100% sure, but I don't think the `.length` bit would be needed. There's a few examples within the repo.

File: packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts [hunk 3]
```diff
@@ -426,6 +430,15 @@ export class StateMachine extends StateMachineBase {
...
+      throw new Error('StateMachine name must match [0-9a-zA-Z\_\-]+');
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:438]:
```suggestion
      throw new Error(`State Machine name must match "[0-9a-zA-Z\_\-]+". received: ${stateMachineName}`);
```

**Review by shivlaks — CHANGES_REQUESTED** [review]
@jonnekaunisto thanks for submitting a fix for this! I've left a couple of minor suggestions.

File: packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts (hunk 2 above)
Comment (jonnekaunisto) [reply packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts:437]: Yeah here's a similar example to what you suggest: https://github.com/aws/aws-cdk/blob/404b5569bb76e3f579dd3005dcf06eeeae9df7f6/packages/%40aws-cdk/aws-route53/lib/util.ts#L26

File: packages/@aws-cdk/aws-stepfunctions/test/state-machine.test.ts [hunk 4]
```diff
@@ -62,6 +62,31 @@ describe('State Machine', () => {
...
+        stateMachineType: stepfunctions.StateMachineType.EXPRESS,
+      });
+    };
+    const tooLongName = 'M'.repeat(81);
+    const invalidCharactersName = '*';
+
+    // THEN
+    expect(() => {
+      createStateMachine(tooLongName);
+    }).toThrow(`State Machine name must be between 1 and 80 characters. Received: ${tooLongName}`);
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions/test/state-machine.test.ts:83]: let's also add one more expect for a blank string to cover the other side of the bounds.

**Review by shivlaks — COMMENTED** [review]
@jonnekaunisto looks great! one minor test case to add and we're good to go here.
thanks for turning it around so quickly!!

**Review by shivlaks — DISMISSED** [review]
left 1 minor comment - add a test (expect statement) for lower bound in length validation.

**Review by shivlaks — APPROVED** [review]
@jonnekaunisto thanks for the contribution!!

**Pull request merged**